[WiP] Adding new backend for MapD

xmnlab opened the pull request:
Also resolves #1418  and resolves #893

cpcloud commented on ibis/expr/api.py:
```diff
@@ -388,6 +389,8 @@ def row_number():
 
 e = ops.E().to_expr()
 
+pi = ops.Pi().to_expr().name('pi')
```
I would leave this unnamed for now. No reason to make this choice for users.

cpcloud commented on ibis/expr/api.py:
```diff
@@ -1300,6 +1314,16 @@ def _integer_to_interval(arg, unit='s'):
+acos = _unary_op('acos', ops.Acos)
+asin = _unary_op('asin', ops.Asin)
+atan = _unary_op('atan', ops.Atan)
+atan2 = _generic_op('atan2', ops.Atan2)
```
This is a binary operation right? There should be something like `_binary_op` function lying around here.

xmnlab replied:
you're right. I think this is the right function: `_binop_expr`

cpcloud commented on ibis/expr/operations.py:
```diff
@@ -516,6 +521,53 @@ class Log10(Logarithm):
 
+# TRIGONOMETRIC OPERATIONS
+
+class TrigonometryUnary(UnaryOp):
```
Can you change the name from `TrigonometryUnary` to `TrigonometricUnary`, and do the same for `TrigonometryBinary`.

cpcloud commented on ibis/expr/operations.py:
```diff
@@ -2183,6 +2235,14 @@ def output_type(self):
 
 
+class Pi(Constant):
+    """
```
I guess you could saying something here like `The constant pi.`

cpcloud commented on ibis/mapd/client.py:
```diff
@@ -0,0 +1,326 @@
+        elif GPUDataFrame is not None and isinstance(
+                self.cursor, GPUDataFrame
+        ):
+            result = self.cursor
```
Since you're executing the same code (assigning `self.cursor` to `result`) in the case that the cursor is a pandas DataFrame or a GPUDataFrame, can you remove the last two `elif`s? Is there a case where `self.cursor is not None` and it's not either a pandas DataFrame or a GPUDataFrame?

xmnlab replied:
you're right! thanks!

cpcloud commented on ibis/mapd/client.py:
```diff
@@ -0,0 +1,326 @@
+    """
+    def _fetch(self, cursor):
+        # check if cursor is a pymapd cursor.Cursor
+        return self.schema().apply_to(cursor.to_df())
```
We should probably come up with a simple interface for getting a `DataFrame` out of a `Cursor`, though you don't need to do that in this PR.

cpcloud replied:
That way you only have to inherit from `Query`.

cpcloud commented on ibis/mapd/client.py:
```diff
@@ -0,0 +1,326 @@
+        :param port:
+        :param dbname:
+        :param protocol:
+        :param execution_type:
```
Please use the [numpy docstring conventions](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt) here.

cpcloud commented on ibis/mapd/client.py:
```diff
@@ -0,0 +1,326 @@
+
+    @property
+    def _table_expr_klass(self):
+        return ir.TableExpr
```
You *should* be able to delete this property. I believe it is provided in the parent class.

cpcloud commented on ibis/mapd/client.py:
```diff
@@ -0,0 +1,326 @@
+        return result
+
+    def _fully_qualified_name(self, name, database):
+        if bool(fully_qualified_re.search(name)):
```
No need to call `bool` on the test of an `if` statement, Python does that behind the scenes.

cpcloud commented on ibis/mapd/client.py:
```diff
@@ -0,0 +1,326 @@
+            return name
+
+        database = database or self.current_database
+        return '{0}.{1}'.format(database, name)
```
Leave out the 0 and 1 in the format string.

cpcloud commented on ibis/mapd/client.py:
```diff
@@ -0,0 +1,326 @@
+        """
+
+        :param table_name:
+        :return:
```
Numpy docstring format

cpcloud commented on ibis/mapd/client.py:
```diff
@@ -0,0 +1,326 @@
+        :return:
+        """
+        database = None
+        table_name_ = table_name.split('.')
```
Instead of string splitting on a dot (which assumes that there are no embedded dots in the table or database name), why not have `_get_table_schema` take a third parameter called `database` whose default value is `None`.

xmnlab replied:
I did it because `table` method  use a qualified name when call `_get_table_schema`

```python
qualified_name = self._fully_qualified_name(name, database)
schema = self._get_table_schema(qualified_name)```

should I change `_fully_qualified_name` instead?

my `_fully_qualified_name` 

```python

    def _fully_qualified_name(self, name, database):
        if fully_qualified_re.search(name):
            return name

        database = database or self.current_database
        return '{}.{}'.format(database, name)

```

cpcloud commented on ibis/mapd/client.py:
```diff
@@ -0,0 +1,326 @@
+        :param query:
+        :return:
+        """
+        if not self.execution_type == 3:
```
1. Don't negate `==` by using `not` at the beginning of the expression, so `not a == b` should be written `a != b`.
1. What does `3` mean here? Can you put it into some human readable enum?

cpcloud commented on ibis/mapd/client.py:
```diff
@@ -0,0 +1,326 @@
+        equal to the current database.
+        """
+        if name == self.current_database or (
+            name is None and name != self.current_database
```
Can `self.current_database` be `None`? If not, then you don't need the second part of the `and` here since `name is None` would imply that `name != self.current_database`.

cpcloud commented on ibis/mapd/client.py:
```diff
@@ -0,0 +1,326 @@
+        -------
+        if_exists : boolean
+        """
+        return len(self.list_tables(like=name, database=database)) > 0
```
You can return `bool(self.list_tables(...))`

cpcloud commented on ibis/mapd/client.py:
```diff
@@ -0,0 +1,326 @@
+            col_names.append(col.name)
+            col_types.append(MapDDataType.parse(col.type))
+
+        return sch.schema(col_names, col_types)
```
This can be more concise if you write it as `sch.schema([(col.name, MapDDataType.parse(col.type)) for col in self.con.get_table_details(table_name)])`

cpcloud replied:
Of course, insert some line breaks as necessary to make sure you pass `flake8`

cpcloud commented on ibis/mapd/compiler.py:
```diff
@@ -0,0 +1,271 @@
+            return None
+
+        lines = []
+        if len(self.group_by) > 0:
```
Just write `self.group_by`.

cpcloud commented on ibis/mapd/compiler.py:
```diff
@@ -0,0 +1,271 @@
+
+        lines = []
+        if len(self.group_by) > 0:
+            columns = ['{0}'.format(expr.get_name())
```
No need for the indexed format string here.

cpcloud commented on ibis/mapd/compiler.py:
```diff
@@ -0,0 +1,271 @@
+        return super().format_select_set()
+
+    def format_group_by(self):
+        if not len(self.group_by):
```
Can just write `not self.group_by`.

cpcloud commented on ibis/mapd/compiler.py:
```diff
@@ -0,0 +1,271 @@
+            clause = 'GROUP BY {0}'.format(', '.join(columns))
+            lines.append(clause)
+
+        if len(self.having) > 0:
```
Just write `if self.having`.

cpcloud commented on ibis/mapd/compiler.py:
```diff
@@ -0,0 +1,271 @@
+        if len(self.group_by) > 0:
+            columns = ['{0}'.format(expr.get_name())
+                       for expr in self.group_by]
+            clause = 'GROUP BY {0}'.format(', '.join(columns))
```
No indexed format string.

cpcloud commented on ibis/mapd/compiler.py:
```diff
@@ -0,0 +1,271 @@
+            for expr in self.having:
+                translated = self._translate(expr)
+                trans_exprs.append(translated)
+            lines.append('HAVING {0}'.format(' AND '.join(trans_exprs)))
```
No indexed format string.

cpcloud commented on ibis/mapd/compiler.py:
```diff
@@ -0,0 +1,271 @@
+        return _name_expr(translated, name)
+
+    @classmethod
+    def compiles(cls, klass, f=None):
```
You should not need to redefine this.

cpcloud commented on ibis/mapd/compiler.py:
```diff
@@ -0,0 +1,271 @@
+    # compile the argument
+    compiled_arg = translator.translate(x)
+
+    return 'STDDEV%s%s(%s)' % (how, f_type, compiled_arg)
```
Use format strings instead of string specs here.

cpcloud commented on ibis/mapd/compiler.py:
```diff
@@ -0,0 +1,271 @@
+    args_size = len(args)
+
+    x = args[0]
+    how = '' if args_size < 2 else '_%s' % args[1].upper()[:4]
```
This is very hard to understand. Can you take a look at one of the other string generating backends that needs to deal with this and see if you can clean up your code based on that?

cpcloud commented on ibis/mapd/compiler.py:
```diff
@@ -0,0 +1,271 @@
+    args_size = len(args)
+
+    x = args[0]
+    how = '' if args_size < 2 else '_%s' % args[1].upper()[:4]
```
See above comment.

cpcloud commented on ibis/mapd/compiler.py:
```diff
@@ -0,0 +1,271 @@
+    # compile the argument
+    compiled_arg = translator.translate(x)
+
+    return '%s%s%s(%s)' % (f_name, how, f_type, compiled_arg)
```
Format strings.

cpcloud commented on ibis/mapd/compiler.py:
```diff
@@ -0,0 +1,271 @@
+    compiled_x = translator.translate(x)
+    compiled_y = translator.translate(y)
+
+    return 'CORR%s(%s, %s)' % (f_type, compiled_x, compiled_y)
```
format strings.

cpcloud commented on ibis/mapd/compiler.py:
```diff
@@ -0,0 +1,271 @@
+    # compile the argument
+    compiled_arg = translator.translate(arg)
+
+    return 'CHAR_LENGTH(%s)' % compiled_arg
```
Format strings.

cpcloud commented on ibis/mapd/tests/test_compiler.py:
```diff
@@ -0,0 +1,14 @@
+import ibis.expr.datatypes as dt
+
+
+def test_timestamp_accepts_date_literals(alltypes):
```
This is a BigQuery specific test. Do you really need it?

cpcloud commented on ibis/mapd/udf.py:
```diff
@@ -0,0 +1,5 @@
+"""
+User Defined Function
```
I wouldn't add this module unless there's support for this in MapD.

cpcloud commented on ibis/mapd/tests/test_client.py:
```diff
@@ -0,0 +1,365 @@
+    assert result == expected
+
+'''
+def test_simple_aggregate_execute(alltypes, df):
```
Many of these tests are BigQuery specific. Can you remove them from here?

The preferred alternative is to add appropriate tests in `ibis/tests/all/test_*.py`. To do that, you'll also need to add a `MapD` class in `ibis/tests/all/backends.py`. There are many examples in that file to get you started.

cpcloud commented on ibis/mapd/operations.py:
```diff
@@ -0,0 +1,798 @@
+from six import StringIO
```
It looks like a lot of the functions and objecs in here are duplicated from either the impala or bigquery backends. Can you see if you can reuse some of their functions so we have only what's needed for MapD?

xmnlab replied:
Ok. I will take a look :) thanks!

xmnlab commented:
When I try to use a mean after a group by the backend return this compile string:

```python
table.group_by('origin_name').dest_lat.mean().compile()
```

```sql
SELECT "origin_name", avg("dest_lat") AS mean(dest_lat)
FROM mapd.flights_2008_10k
GROUP BY origin_name
```

if I try to put a name to the expression, the backend raise an error:

```python
> > >  table.group_by('origin_name').dest_lat.mean().name('new_name').compile()

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-38-25574603f8fa> in <module>()
      1 # TODO: resolve aggregation name
----> 2 table.group_by('origin_name').dest_lat.mean().name('tst').compile()

/media/xmn/29bd0df7-9ccb-4dd1-93a0-e47e2b4f2fc2/dev/quansight/ibis/ibis/expr/types.py in __getattr__(self, key)
    441 
    442         if key not in schema:
--> 443             raise AttributeError(key)
    444 
    445         try:

AttributeError: name
```

How can I fixed that? the default name breaks on my backend.

@cpcloud do you have any idea?

xmnlab commented:
@cpcloud 

we have a special case of data type that using encoding key word. Example:

```sql

CREATE TABLE IF NOT EXISTS tweets (
  tweet_id BIGINT NOT NULL,
  tweet_time TIMESTAMP NOT NULL ENCODING FIXED(32),
  lat REAL,
  lon REAL,
  sender_id BIGINT NOT NULL,
  sender_name TEXT NOT NULL ENCODING DICT,
  location TEXT ENCODING DICT,
  source TEXT ENCODING DICT,
  reply_to_user_id BIGINT,
  reply_to_tweet_id BIGINT,
  lang TEXT ENCODING DICT,
  followers INT,
  followees INT,
  tweet_count INT,
  join_time TIMESTAMP ENCODING FIXED(32),
  tweet_text TEXT,
  state TEXT ENCODING DICT,
  county TEXT ENCODING DICT,
  place_name TEXT,
  state_abbr TEXT ENCODING DICT,
  county_state TEXT ENCODING DICT,
  origin TEXT ENCODING DICT);

```

What is the best approach to do it?

Refs:

https://www.mapd.com/docs/latest/mapd-core-guide/fixed-encoding/
https://www.mapd.com/docs/latest/mapd-core-guide/tables/

xmnlab commented:
@cpcloud 

How is the best way to rewrite ibis.year()?

for our backend, the value returned should be quoted:

```sql
INTERVAL SELECT NOW() - INTERVAL '1' YEAR AS tmp
```

cpcloud commented:
@xmnlab I should be able to review this today or tomorrow

xmnlab commented:
@cpcloud thanks!

This is still a WIP .. there is something more I need to add. But any orientation would be very helpful

xmnlab commented:
to add tests to mapd backend, should I add this the this option on `ci` folder?

kszucs commented on ibis/expr/api.py:
```diff
@@ -1292,20 +1295,38 @@ def _integer_to_interval(arg, unit='s'):
 log10 = _unary_op('log10', ops.Log10)
 ln = _unary_op('ln', ops.Ln)
+radians = _unary_op('radians', ops.Radians)
+rad2deg = _unary_op('degrees', ops.Degrees)
```
Due to `rad2deg` and `deg2rad` are merely aliases for `degrees` and `radians` You don't need to define them again, just add to `_numeric_value_methods`:

```python
_numeric_value_methods = dict(
    degrees=degrees,
    radians=radians,
    rad2deg=degrees,
    deg2rad=radians
)
```

kszucs commented on ibis/expr/api.py:
```diff
@@ -3110,3 +3203,11 @@ def _table_drop(self, fields):
+# geometric operation
+
+def distance(from_lon, from_lat, to_lon, to_lat):
+    """
```
Looks like missing a little docstring :)

kszucs commented on ibis/expr/operations.py:
```diff
@@ -284,6 +284,11 @@ class UnaryOp(ValueOp):
 
+class BinaryOp(ValueOp):
+    left = Arg(rlz.any)
+    right = Arg(rlz.any)
```
There already exists a [BinaryOp](https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations.py#L702). 

However I think this is the proper implementation for `BinaryOp` and the old one should be removed. If I recall correctly [Comparison](https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations.py#L2048) is the only operation which depends on calling `_maybe_cast_args`. 

So if You move [BinaryOp.__init__](https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations.py#L711) to [Comparison](https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations.py#L2048) along with the [comments](https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations.py#L705) then We could delete the [old BinaryOp](https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations.py#L702) and use the one defined here instead.

kszucs commented on ibis/expr/operations.py:
```diff
@@ -804,6 +872,38 @@ class Variance(VarianceBase):
 
 
+class Correlation(Reduction):
+    """
```
Formatting.

kszucs commented on ibis/expr/operations.py:
```diff
@@ -804,6 +872,38 @@ class Variance(VarianceBase):
+            dtype = self.left.type().largest
+        else:
+            dtype = dt.float64
+        return dtype.scalar_type()
```
@cpcloud  I'm not sure about the type promotion here, do We want decimals here?

kszucs replied:
There is also [rules.numeric_like](https://github.com/ibis-project/ibis/blob/master/ibis/expr/rules.py#L370) which we might want to extend.

cpcloud replied:
@xmnlab You can leave out decimal types for now and if/when we need support for them we can add it. We can also add it in a follow up PR before that as well.

xmnlab replied:
great! so should I keep that, right?

cpcloud replied:
No, leave them out for now

kszucs commented on ibis/expr/operations.py:
```diff
@@ -804,6 +872,38 @@ class Variance(VarianceBase):
+    how = Arg(rlz.isin({'sample', 'pop'}), default=None)
+    where = Arg(rlz.boolean, default=None)
+
+    def output_type(self):
```
@cpcloud The same as above.

cpcloud replied:
Yes @xmnlab leave out DecimalValue stuff for now

kszucs commented on ibis/expr/operations.py:
```diff
@@ -516,6 +521,65 @@ class Log10(Logarithm):
+class TrigonometricUnary(UnaryOp):
+    """Trigonometric base unary"""
+    arg = Arg(rlz.numeric)
+    output_type = rlz.shape_like('arg', 'float')
```
For consistency please use `dt.float` instead of `'float'`.

kszucs commented on ibis/expr/operations.py:
```diff
@@ -516,6 +521,65 @@ class Log10(Logarithm):
+    """Trigonometric base binary"""
+    left = Arg(rlz.numeric)
+    right = Arg(rlz.numeric)
+    output_type = rlz.shape_like('left', 'float')
```
I think if any of the operands are columnar then we need to return columnar result too. 
To achieve that use:

```python
output_type = rlz.shape_like('args', dt.float64)
```

kszucs commented on ibis/expr/operations.py:
```diff
@@ -2692,3 +2800,17 @@ def root_tables(self):
+    from_lat = Arg(rlz.column(rlz.numeric))
+    to_lon = Arg(rlz.column(rlz.numeric))
+    to_lat = Arg(rlz.column(rlz.numeric))
+    output_type = rlz.shape_like('from_lon', dt.float)
```
@xmnlab Can't be any of the arguments scalar like?

xmnlab replied:
@kszucs yes any of the args could be a scalar or a column, should I changed to just `rlz.numeric` ?

kszucs replied:
@xmnlab Then use the following definition:

```python
 from_lat = Arg(rlz.numeric)
 to_lon = Arg(rlz.numeric)
 to_lat = Arg(rlz.numeric)
 output_type = rlz.shape_like('args', dt.float64)
```

This way, if any of the arguments is column like, then the return type will be columnar shaped too.

kszucs commented on ibis/mapd/compiler.py:
```diff
@@ -0,0 +1,220 @@
+
+_add_methods(
+    ir.StringValue, dict(
+        byte_length=_unary_op('length', mapd_ops.ByteLength)
```
I think this should be moved to `ibis/expr/api.py`

cpcloud replied:
Indeed, having a different API for bytes vs strings isn't great, but one may also want to compute the number of bytes in a string value as well as the semantic length of the string in human-readable characters

kszucs commented on ibis/mapd/operations.py:
```diff
@@ -0,0 +1,644 @@
+import ibis.expr.types as ir
+import ibis.expr.operations as ops
+
+_mapd_unit_names = {
```
[resolution](https://github.com/ibis-project/ibis/blob/master/ibis/expr/datatypes.py#L416) property might be used for this

kszucs commented on ibis/mapd/operations.py:
```diff
@@ -0,0 +1,644 @@
+    )
+
+
+def _add_methods(dtype, function_ops, forced=False):
```
@xmnlab Are these used?

xmnlab replied:
it is not being used anymore .. thanks!

kszucs commented on ibis/mapd/operations.py:
```diff
@@ -0,0 +1,644 @@
+
+# AGGREGATION
+
+class ApproxCountDistinct(ops.Reduction):
```
This should be moved to `ibis.expr.api`

xmnlab replied:
`ApproxCountDistinct` is a alternative to` HLLCardinality` (https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations.py#L819) ... the difference is ApproxCountDistinct receive a `int` parameter.

Some info about this on MapD:

> APPROX_COUNT_DISTINCT(x, e) gives an approximate count of the value x, based on an expected error rate defined in e. The error rate is an integer value from 1 to 100. The lower the value of e, the higher the precision, and the higher the memory cost. Select a value for e based on the level of precision required. On large tables with large cardinalities, consider using APPROX_COUNT_DISTINCT when possible to preserve memory.

kszucs replied:
I think mapD should compile `HLLCardinality` operation to `APPROX_COUNT_DISTINCT` with a default error rate for now. Later we can introduce a separate operation or the error argument to `HLLCardinality` itself, but that needs input argument validation for other backends because they don't support other than `error=None`.

@cpcloud What do You think?

kszucs commented on ibis/mapd/operations.py:
```diff
@@ -0,0 +1,644 @@
+# classes
+
+# MATH
+class Log(ops.Ln):
```
Seems unused.

kszucs commented:
@xmnlab Could You please add `MapD` as a [backend](https://github.com/ibis-project/ibis/blob/master/ibis/tests/backends.py)? 

We have a couple of (mostly) backend agnostic [tests cases](https://github.com/ibis-project/ibis/tree/master/ibis/tests/all) so check common operations.

kszucs commented:
In general this PR looks great, mostly just need to increase the test coverage. You've already passed the hard parts of ibis' learning curve :)

xmnlab commented:
@kszucs 

for unsupported operations should I do something like:

```python

_unsupported_ops = [
    ops.WindowOp,
    ops.DecimalPrecision,
    ops.DecimalScale,
    ops.BaseConvert,
    ops.CumulativeSum,
    ops.CumulativeMin,
    ops.CumulativeMax,
    ops.CumulativeMean,
    ops.CumulativeAny,
    ops.CumulativeAll,
    ops.IdenticalTo,

    ops.RowNumber,
    ops.DenseRank,
    ops.MinRank,
    ops.PercentRank,

    ops.FirstValue,
    ops.LastValue,
    ops.NthValue,
    ops.Lag,
    ops.Lead,
    ops.NTile
]
_unsupported_ops = {k: raise_error for k in _unsupported_ops}

_operation_registry.update(_unsupported_ops)

```

xmnlab commented:
or just should I use `additional_skipped_operations = frozenset({ ... })` on `MapD` `Backend` class?

kszucs commented:
@xmnlab For now use `_unsupported_ops`

kszucs commented:
@xmnlab You should also add `MapD` to the list of supported backends in the top level README - flagged as experimental.

kszucs commented:
Hey @xmnlab ! Let me know when I could review again, I'd like to see this PR merged :)

xmnlab commented:
@kszucs 

do you know why this is happing?:

```
raceback (most recent call last):
  File "/ibis/ci/feedstock.py", line 12, in <module>
    import ibis
  File "/ibis/ibis/__init__.py", line 21, in <module>
    import ibis.expr.api as api
  File "/ibis/ibis/expr/api.py", line 392, in <module>
    pi = ops.Pi().to_expr()
  File "/ibis/ibis/expr/operations.py", line 112, in to_expr
    self._expr_cached = self._make_expr()
  File "/ibis/ibis/expr/operations.py", line 116, in _make_expr
    klass = self.output_type()
  File "/ibis/ibis/expr/operations.py", line 2314, in output_type
    return partial(ir.FloatingScalar, dtype=dt.float64)
NameError: name 'partial' is not defined
Exited with code 1

```

xmnlab commented:
`partial` was removed from operations. I included that again.

kszucs commented:
I was about to respond. You are [correct](https://github.com/ibis-project/ibis/commit/3c07ebf56f102892978eee75c0dfb351ee290bc6), use `functools.partial` instead.

xmnlab commented:
@kszucs 

with py2 there is a issue related with thrift version ... 
there is some reason to use `thrift<=0.9.3'` ? for mapd backend on python2 env it should be necessary thrift 0.10.0 (same problem https://github.com/Ericsson/codechecker/issues/546#issuecomment-331707747)

kszucs commented:
@cpcloud probably can answer it

xmnlab commented:
@kszucs 

It seems mapd core is working on python27_test (until data import step) .. but it is failing with impala:

```python
AttributeError: 'TBufferedTransport' object has no attribute 'trans'
/tmp/ibis-testing-data/awards_players.csv
```

What should be the best approach in this case? could we disable mapd for python2?

on python36_test, it raised this 5 errors:

```python
Your build ran 6927 tests with 5 failures:

1. test_cross_project_query - ibis.bigquery.tests.test_client ibis/bigquery/tests/test_client.py
> ibis/bigquery/tests/test_client.py:406: in test_cross_project_query

2. test_repeated_project_name - ibis.bigquery.tests.test_client ibis/bigquery/tests/test_client.py
> ibis/bigquery/tests/test_client.py:460: in test_repeated_project_name

3. test_set_database - ibis.bigquery.tests.test_client ibis/bigquery/tests/test_client.py
> ibis/bigquery/tests/test_client.py:427: in test_set_database

4. test_any_all - ibis.impala.tests.test_exprs.TestValueExprs ibis/impala/tests/test_exprs.py
> AssertionError: assert 'sum(CASE WHE...E 0 END) >= 0' == 'sum(`f` = 0) > 0'   - sum(CASE WHEN `f` = 0 THEN 1 ELSE 0 END) >= 0   + sum(`f` = 0) > 0

5. test_window[Impala-cumany] - ibis.tests.all.test_window ibis/tests/all/test_window.py
> AssertionError: Series are different  Series values are different (90.0 %) [left]:  [True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, ...] [right]: [False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, ...]

```

do you have any idea why impala seems to be affected?

cpcloud commented on ibis/tests/all/test_string.py:
```diff
@@ -112,7 +112,8 @@ def test_string_col_is_unicode(backend, alltypes, df):
             lambda t: t.string_col.str.len().astype('int32'),
-            id='length'
+            id='length',
+            marks=pytest.mark.xfail,
```
This cannot be marked `xfail`. Does mapd not implement this operation?

xmnlab replied:
At this moment mapd doesn't support thinks like "SELECT 1;" or "SELECT COS(1);" in other word without the table identification.

cpcloud commented on ibis/tests/all/test_numeric.py:
```diff
@@ -54,25 +54,79 @@ def test_isnan_isinf(backend, con, alltypes, df,
+    ),
+    param(
+        L(11) % 3, 11 % 3,
+        id='mod', marks=pytest.mark.xfail
```
Please revert making these `xfail`. It is import that these actually pass (and not xpass)

xmnlab replied:
At this moment mapd doesn't support thinks like "SELECT 1;" or "SELECT COS(1);" in other word without the table identification.

how could I skip that for mapd?

cpcloud commented on ibis/tests/all/test_aggregation.py:
```diff
@@ -72,21 +72,25 @@
         lambda t, where: t.double_col.var(how='pop'),
         lambda t, where: t.double_col.var(ddof=0),
         id='double_col_var_pop',
+        marks=pytest.mark.xfail,
```
These all definitely need to pass

xmnlab replied:
I have no idea why but the mapd and pandas results are different. 
that is why put that.
should I put these operation as unsupported to move on?

xmnlab replied:
I found the problem. I will fix that.

cpcloud commented on ibis/mapd/tests/conftest.py:
```diff
@@ -0,0 +1,56 @@
+def translate():
+    """
+
+    :return:
```
You don't really need a docstring here.

xmnlab replied:
ok. thanks

cpcloud commented on ibis/expr/operations.py:
```diff
@@ -2694,3 +2799,17 @@ def output_type(self):
+
+class Distance(ValueOp):
+    """
+    Calculates distance in meters between two WGS-84 positions.
```
What is WGS-84? Does it really need to be in the docstring here?

xmnlab replied:
> WGS84 is the standard U.S. Department of Defense definition of a global reference system for geospatial information and is the reference system for the Global Positioning System (GPS). 

I can remove if you want it is unnecessary

cpcloud replied:
What sort of constraints does this put on the operation? This is different than the standard distance calculation, right?

cpcloud replied:
Hm, I'm not comfortable adding a distance function that assumes what appears to be a domain specific coordinate system. Can we make this Euclidean distance instead and follow up after this PR about adding the WGS version?

xmnlab replied:
Yes, we can do that. We just cannot test it because mapd doesn't have a Euclidean operation. 
I am just not sure if common Euclidean distance function works with lat lon parameters .. so maybe would be better to remove this function and create a issue to follow this discussion.

cpcloud commented on ibis/expr/operations.py:
```diff
@@ -516,6 +523,65 @@ class Log10(Logarithm):
+class Radians(UnaryOp):
+    """Converts radians to degrees"""
+    arg = Arg(rlz.floating)
+    output_type = rlz.shape_like('arg', dt.float)
```
These are definitely float32 and not double?

cpcloud replied:
The output type should be the same as the input type, so if it's a `float32` return a `float32` otherwise return a `float64`.

xmnlab replied:
ok .. I changed to 

```
output_type = rlz.shape_like('arg', dtype=arg.type())
```

xmnlab replied:
sorry, I changed to:

```
    def output_type(self):
        return rlz.shape_like('arg', dtype=self.arg.type())
```

xmnlab replied:
Actually .. I think the input could be defined as numeric

> The output type should be the same as the input type, so if it's a float32 return a float32 otherwise return a float64.

could you point me some example to do that?

or could we just put as float64 and treat different output (if necessary) in a follow up PR?

cpcloud commented on ibis/expr/api.py:
```diff
@@ -1419,6 +1447,38 @@ def variance(arg, where=None, how='sample'):
+
+def covariance(left, right, where=None, how='sample'):
+    """
+    Compute covariance of numeric array
```
This should be *two* numeric arrays.

cpcloud commented on ibis/expr/api.py:
```diff
@@ -1419,6 +1447,38 @@ def variance(arg, where=None, how='sample'):
 
+def correlation(left, right, where=None, how='sample'):
+    """
+    Compute corralation of numeric array
```
This should be two numeric arrays. Also, correlation not corralation

cpcloud commented on ibis/expr/api.py:
```diff
@@ -3108,3 +3199,26 @@ def _table_drop(self, fields):
+def distance(from_lon, from_lat, to_lon, to_lat):
+    """
+    Distance between origin longitude and latitude and
+    destine longitude and latitude
```
It looks like `destine` is actually an English word (which I did not know), but I wonder if this reads more clearly as "destination" instead

cpcloud commented on appveyor.yml:
```diff
@@ -49,9 +55,9 @@ test_script:
+  - "%CONDA% install --channel conda-forge pytables numpy sqlalchemy psycopg2 pymapd graphviz click mock plumbum flake8 pytest"
   - "%CONDA% list"
-  - "pip install -e .\"[sqlite, postgres, mysql, visualization, pandas, csv, hdf5]\""
+  - "pip install -e .\"[sqlite, postgres, mysql, visualization, pandas, csv, hdf5, mapd]\""
```
We probably do not want to test `mapd` on Windows since we don't have a way to hit a mapd database on CI without docker, and even if appveyor has docker and there's a mapd-for-windows docker image we should set that up in a follow up PR.

xmnlab replied:
ok!

cpcloud commented:
@xmnlab looks like this is failing due to a bug in one of our bigquery tests, i'll have a fix for this up shortly

cpcloud commented:
@xmnlab Rebase on top of master to pick up #1445 which will allow the bigquery tests to be skipped on your fork

xmnlab commented:
@cpcloud thanks!!

xmnlab commented:
@cpcloud @kszucs 

At this moment I see, at least, 2 import points:

- make unavailable ibis.mapd for python2 (about thrift conflict)
- Skip tests for ibis.mapd for some numeric tests:

```
param(L(-5).abs(), 5, id='abs-neg'),
param(L(5).abs(), 5, id='abs'),
param(ibis.least(L(10), L(1)), 1, id='least'),
param(ibis.greatest(L(10), L(1)), 10, id='greatest'),
param(L(5.5).round(), 6.0, id='round'),
param(L(5.556).round(2), 5.56, id='round-digits'),
param(L(5.556).ceil(), 6.0, id='ceil'),
param(L(5.556).floor(), 5.0, id='floor'),
param(L(5.556).exp(), math.exp(5.556), id='expr'),
param(L(5.556).sign(), 1, id='sign-pos'),
param(L(-5.556).sign(), -1, id='sign-neg'),
param(L(0).sign(), 0, id='sign-zero'),
param(L(5.556).sqrt(), math.sqrt(5.556), id='sqrt'),
param(L(5.556).log(2), math.log(5.556, 2), id='log-base'),
param(L(5.556).ln(), math.log(5.556), id='ln'),
param(L(5.556).log2(), math.log(5.556, 2), id='log2'),
param(L(5.556).log10(), math.log10(5.556), id='log10'),
param(L(11) % 3, 11 % 3, id='mod'),
```

I need to skip this test because MapD doesn't support query with no table (`Exception: Non-empty LogicalValues not supported yet`).

I tried to use: 

```
additional_skipped_operations = frozenset({
        ops.Abs, ops.Round, ops.Ceil, ops.Floor, ops.Exp, ops.Sign, ops.Sqrt,
        ops.Ln, ops.Log10, ops.Modulus
})
```

but it seems not suppported yet.

xmnlab commented:
@kszucs 

there is some  problem on sqlite import that I could understand why ... do you have any idea what could be the problem?

kszucs commented:
@cpcloud has recently changed that -> ping

xmnlab commented:
@kszucs 

for the CI tests, I fixed pandas <0.23.

for python 2 tests this is the result:

`4 failed, 5709 passed, 757 skipped, 9 deselected, 231 xfailed, 70 xpassed, 9 warnings in 427.40 seconds `

```
=================================== FAILURES ===================================
___________________________ test_cross_project_query ___________________________
ibis/bigquery/tests/test_client.py:406: in test_cross_project_query
    dataset_id='bigquery-public-data.stackoverflow')
ibis/bigquery/api.py:52: in connect
    return BigQueryClient(project_id, dataset_id)
ibis/bigquery/client.py:335: in __init__
    self.client = bq.Client(project=self.data_project)
/opt/conda/envs/ibis-dev-2.7/lib/python2.7/site-packages/google/cloud/bigquery/client.py:121: in __init__
    project=project, credentials=credentials, _http=_http)
/opt/conda/envs/ibis-dev-2.7/lib/python2.7/site-packages/google/cloud/client.py:216: in __init__
    Client.__init__(self, credentials=credentials, _http=_http)
/opt/conda/envs/ibis-dev-2.7/lib/python2.7/site-packages/google/cloud/client.py:129: in __init__
    credentials, _ = google.auth.default()
/opt/conda/envs/ibis-dev-2.7/lib/python2.7/site-packages/google/auth/_default.py:271: in default
    credentials, project_id = checker()
/opt/conda/envs/ibis-dev-2.7/lib/python2.7/site-packages/google/auth/_default.py:142: in _get_explicit_environ_credentials
    os.environ[environment_vars.CREDENTIALS])
/opt/conda/envs/ibis-dev-2.7/lib/python2.7/site-packages/google/auth/_default.py:67: in _load_credentials_from_file
    'File {} was not found.'.format(filename))
E   DefaultCredentialsError: File /tmp/gcloud-service-key.json was not found.
__________________________ test_repeated_project_name __________________________
ibis/bigquery/tests/test_client.py:459: in test_repeated_project_name
    project_id='ibis-gbq', dataset_id='ibis-gbq.testing')
ibis/bigquery/api.py:52: in connect
    return BigQueryClient(project_id, dataset_id)
ibis/bigquery/client.py:335: in __init__
    self.client = bq.Client(project=self.data_project)
/opt/conda/envs/ibis-dev-2.7/lib/python2.7/site-packages/google/cloud/bigquery/client.py:121: in __init__
    project=project, credentials=credentials, _http=_http)
/opt/conda/envs/ibis-dev-2.7/lib/python2.7/site-packages/google/cloud/client.py:216: in __init__
    Client.__init__(self, credentials=credentials, _http=_http)
/opt/conda/envs/ibis-dev-2.7/lib/python2.7/site-packages/google/cloud/client.py:129: in __init__
    credentials, _ = google.auth.default()
/opt/conda/envs/ibis-dev-2.7/lib/python2.7/site-packages/google/auth/_default.py:271: in default
    credentials, project_id = checker()
/opt/conda/envs/ibis-dev-2.7/lib/python2.7/site-packages/google/auth/_default.py:142: in _get_explicit_environ_credentials
    os.environ[environment_vars.CREDENTIALS])
/opt/conda/envs/ibis-dev-2.7/lib/python2.7/site-packages/google/auth/_default.py:67: in _load_credentials_from_file
    'File {} was not found.'.format(filename))
E   DefaultCredentialsError: File /tmp/gcloud-service-key.json was not found.
_________________________ TestValueExprs.test_any_all __________________________
ibis/impala/tests/test_exprs.py:348: in test_any_all
    self._check_expr_cases(cases)
ibis/impala/tests/test_exprs.py:38: in _check_expr_cases
    assert result == expected
E   AssertionError: assert 'sum(CASE WHE...E 0 END) >= 0' == 'sum(`f` = 0) > 0'
E     - sum(CASE WHEN `f` = 0 THEN 1 ELSE 0 END) >= 0
E     + sum(`f` = 0) > 0
__________________________ test_window[Impala-cumany] __________________________
ibis/tests/all/test_window.py:150: in test_window
    backend.assert_series_equal(left, right)
ibis/tests/backends.py:110: in assert_series_equal
    left, right, *args, **kwargs)
ibis/tests/backends.py:65: in assert_series_equal
    tm.assert_series_equal(*args, **kwargs)
/opt/conda/envs/ibis-dev-2.7/lib/python2.7/site-packages/pandas/util/testing.py:1276: in assert_series_equal
    obj='{obj}'.format(obj=obj))
pandas/_libs/testing.pyx:59: in pandas._libs.testing.assert_almost_equal
    ???
pandas/_libs/testing.pyx:173: in pandas._libs.testing.assert_almost_equal
    ???
/opt/conda/envs/ibis-dev-2.7/lib/python2.7/site-packages/pandas/util/testing.py:1093: in raise_assert_detail
    raise AssertionError(msg)
E   AssertionError: Series are different
E   
E   Series values are different (90.0 %)
E   [left]:  [True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, ...]
E   [right]: [False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, ...]

``` 

for python 3, It seems that the tests are having problems in the connection to the mapd database ...

xmnlab commented:
@kszucs @cpcloud 

it seems there is no issues now about mapd tests on CI. 
I needed to remove load_data_tests, because it was broken mapd container.

about python36 tests, these are the failures messages:

```python
=================================== FAILURES ===================================
___________________________ test_cross_project_query ___________________________
ibis/bigquery/tests/test_client.py:406: in test_cross_project_query
    dataset_id='bigquery-public-data.stackoverflow')
ibis/bigquery/api.py:52: in connect
    return BigQueryClient(project_id, dataset_id)
ibis/bigquery/client.py:335: in __init__
    self.client = bq.Client(project=self.data_project)
/opt/conda/envs/ibis-dev-3.6/lib/python3.6/site-packages/google/cloud/bigquery/client.py:121: in __init__
    project=project, credentials=credentials, _http=_http)
/opt/conda/envs/ibis-dev-3.6/lib/python3.6/site-packages/google/cloud/client.py:216: in __init__
    Client.__init__(self, credentials=credentials, _http=_http)
/opt/conda/envs/ibis-dev-3.6/lib/python3.6/site-packages/google/cloud/client.py:129: in __init__
    credentials, _ = google.auth.default()
/opt/conda/envs/ibis-dev-3.6/lib/python3.6/site-packages/google/auth/_default.py:271: in default
    credentials, project_id = checker()
/opt/conda/envs/ibis-dev-3.6/lib/python3.6/site-packages/google/auth/_default.py:142: in _get_explicit_environ_credentials
    os.environ[environment_vars.CREDENTIALS])
/opt/conda/envs/ibis-dev-3.6/lib/python3.6/site-packages/google/auth/_default.py:67: in _load_credentials_from_file
    'File {} was not found.'.format(filename))
E   google.auth.exceptions.DefaultCredentialsError: File /tmp/gcloud-service-key.json was not found.
__________________________ test_repeated_project_name __________________________
ibis/bigquery/tests/test_client.py:459: in test_repeated_project_name
    project_id='ibis-gbq', dataset_id='ibis-gbq.testing')
ibis/bigquery/api.py:52: in connect
    return BigQueryClient(project_id, dataset_id)
ibis/bigquery/client.py:335: in __init__
    self.client = bq.Client(project=self.data_project)
/opt/conda/envs/ibis-dev-3.6/lib/python3.6/site-packages/google/cloud/bigquery/client.py:121: in __init__
    project=project, credentials=credentials, _http=_http)
/opt/conda/envs/ibis-dev-3.6/lib/python3.6/site-packages/google/cloud/client.py:216: in __init__
    Client.__init__(self, credentials=credentials, _http=_http)
/opt/conda/envs/ibis-dev-3.6/lib/python3.6/site-packages/google/cloud/client.py:129: in __init__
    credentials, _ = google.auth.default()
/opt/conda/envs/ibis-dev-3.6/lib/python3.6/site-packages/google/auth/_default.py:271: in default
    credentials, project_id = checker()
/opt/conda/envs/ibis-dev-3.6/lib/python3.6/site-packages/google/auth/_default.py:142: in _get_explicit_environ_credentials
    os.environ[environment_vars.CREDENTIALS])
/opt/conda/envs/ibis-dev-3.6/lib/python3.6/site-packages/google/auth/_default.py:67: in _load_credentials_from_file
    'File {} was not found.'.format(filename))
E   google.auth.exceptions.DefaultCredentialsError: File /tmp/gcloud-service-key.json was not found.
_________________________ TestValueExprs.test_any_all __________________________
ibis/impala/tests/test_exprs.py:348: in test_any_all
    self._check_expr_cases(cases)
ibis/impala/tests/test_exprs.py:38: in _check_expr_cases
    assert result == expected
E   AssertionError: assert 'sum(CASE WHE...E 0 END) >= 0' == 'sum(`f` = 0) > 0'
E     - sum(CASE WHEN `f` = 0 THEN 1 ELSE 0 END) >= 0
E     + sum(`f` = 0) > 0
__________________________ test_window[Impala-cumany] __________________________
ibis/tests/all/test_window.py:150: in test_window
    backend.assert_series_equal(left, right)
ibis/tests/backends.py:110: in assert_series_equal
    left, right, *args, **kwargs)
ibis/tests/backends.py:65: in assert_series_equal
    tm.assert_series_equal(*args, **kwargs)
/opt/conda/envs/ibis-dev-3.6/lib/python3.6/site-packages/pandas/util/testing.py:1276: in assert_series_equal
    obj='{obj}'.format(obj=obj))
pandas/_libs/testing.pyx:59: in pandas._libs.testing.assert_almost_equal
    ???
pandas/_libs/testing.pyx:173: in pandas._libs.testing.assert_almost_equal
    ???
/opt/conda/envs/ibis-dev-3.6/lib/python3.6/site-packages/pandas/util/testing.py:1093: in raise_assert_detail
    raise AssertionError(msg)
E   AssertionError: Series are different
E   
E   Series values are different (90.0 %)
E   [left]:  [True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, ...]
E   [right]: [False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, ...]
```

xmnlab commented:
@kszucs 

MapD has a rowid operations and I found row_number() operation on ibis. 

row_number is a window operation ... but I think for mapd, it could be a normal operation. This is an example using rowid:

`select rowid, * from diamonds limit 10;`

what should be the best approach to do that?

kszucs commented:
@xmnlab `RowID` operation should look like `Time` or `Date`, see [here](https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations.py#L2361), but with `int` output. 

On the naming we should look around at other backends too. How about creating a ticket for it and implementing in a follow up PR?

kszucs commented:
@xmnlab Do the tests pass locally?

xmnlab commented:
@kszucs 

great! I will create a issue about rowid, thanks!

about the tests ... the tests locally broken here ... it seems impala is breaking for some reason and all tests fail

xmnlab commented:
@kszucs 

this is the log file from my local test: [ibis_test.log](https://github.com/ibis-project/ibis/files/2063296/ibis_test.log)

xmnlab commented:
@kszucs do you have any idea why these errors are happening?

cpcloud commented:
@xmnlab rebase on master and you'll be able to get past any failures that aren't related this PR.

xmnlab commented:
@kszucs 

just on python35_test there is an internal error ... it seems to be a problem with mapd server

xmnlab commented:
@kszucs @cpcloud 

about the tests locally on python35 env ... here it raises a different errors .. it is the same local errors before related with hiveserver2 and impala ... 

is this PR OK? or what should the next step?

xmnlab commented:
@cpcloud @kszucs 

finally all green :+1:

cpcloud commented:
Ok I'm reviewing this now.

cpcloud commented on ci/requirements-dev-3.5.yml:
```diff
@@ -18,7 +18,9 @@ dependencies:
   - pyarrow>=0.6.0
+  - pymapd>=0.3.2
   - pymysql
+  - pytables
```
Is this a hard dependency of mapd or pymapd?

xmnlab replied:
I just copied from 3.6.yml to have the same environment ... should I remove that?

cpcloud commented on ci/requirements-dev-3.5.yml:
```diff
@@ -29,6 +31,7 @@ dependencies:
   - ruamel.yaml
   - six
   - sqlalchemy>=1.0.0,<1.1.15
+  - thrift
```
ame question as above

xmnlab replied:
I can try to remove that .. but I think pymad was breaking without thrift

cpcloud commented on ci/requirements-docs-3.6.yml:
```diff
@@ -22,6 +22,7 @@ dependencies:
   - plumbum
   - psycopg2
   - pyarrow>=0.6.0
+  - pymapd
```
does this require a version constraint?

xmnlab replied:
you're right! it is better to pin the version here. thanks!

cpcloud commented on ibis/__init__.py:
```diff
@@ -71,6 +72,12 @@
 
+with suppress(ImportError):
+    # pip install ibis-framework[mapd]
+    if sys.version_info[0] < 3:
```
Use `sys.version_info.major` here

xmnlab replied:
ok! thanks

cpcloud commented on ibis/__init__.py:
```diff
@@ -71,6 +72,12 @@
+with suppress(ImportError):
+    # pip install ibis-framework[mapd]
+    if sys.version_info[0] < 3:
+        raise ImportError('ibis.mapd is not allowed it for Python 2.')
```
typo here, should read `ibis.mapd is not allowed for Python 2` or `The mapd backend is not supported under Python 2`

xmnlab replied:
thanks!

xmnlab commented:
@cpcloud thanks a lot for reviewing this PR.

this is a compilation of the main fixes here:

1. ci/requirements-dev-3.5.yml: I just made a rollback and works good.
2. Degrees and Radians: I changed the input to numeric and I changed to output to float64, I also add tests for that on mapd tests.
3. Correlation and Covariance: Sorry I misunderstood that, thanks for the patience :) ... I changed the output to dt.float64.scalar_type() .. I also added tests for these operations on mapd tests.
4. Distance: I am not sure if Euclidean distance common functions work with lat lon .. so I removed that from this PR and I will create now a issue to follow up this. 

if these points are ok for you, I think it is ready for a new review. I just changed the left and right parameters from correlation to column numeric.

Again, thank you so much for your attention.

xmnlab commented:
hi @cpcloud @kszucs 

any update about this PR?

thanks a lot!

xmnlab commented:
there is a conflict now .. I will rebase  here now.

cpcloud commented on ibis/tests/all/test_temporal.py:
```diff
@@ -41,9 +41,11 @@ def test_timestamp_extract(backend, alltypes, df, attr):
     param('W', marks=pytest.mark.xfail),
-    'h', 'm', 's', 'ms', 'us', 'ns'
+    'h', 'm', 's', 'ms', 'us',
+    param('ns', marks=pytest.mark.xfail)
```
Does this pass now that you added the `skipif_backend`?

xmnlab replied:
Yes, for some reason when it try to test `ns` unit .. raise an error that the pytest breaks ... 
I tried using pytest mark for `ns` and it doesn't work .. I need to skip that for mapd.
for another tests here .. I could remove `skipif_backend` and just put some xfail for some units and works good.

xmnlab replied:
you're right! sorry, I will remove this now.

cpcloud commented on ibis/tests/all/conftest.py:
```diff
@@ -1,14 +1,30 @@
+    pytest.param(Impala, marks=pytest.mark.impala)
+]
+
+if sys.version_info.major == 3:
```
should be `> 2`

cpcloud commented on ibis/mapd/tests/test_operations.py:
```diff
@@ -0,0 +1,35 @@
+    ),
+    param(
+        lambda t: t.double_col.cov(t.float_col),
+        91.67005567565313,
```
Ultimately we will want to change these to use a pandas call or numpy call, so that we don't have to depend on hard coded values.

This is fine for now though.

cpcloud commented on ibis/mapd/ddl.py:
```diff
@@ -0,0 +1,422 @@
+from ibis.sql.compiler import DDL, DML
```
Do you have tests for the classes in this file? If not, please add them in a follow-up PR.

cpcloud commented on ibis/expr/api.py:
```diff
@@ -1752,6 +1814,34 @@ def _string_like(self, patterns):
     )
 
 
+def _string_ilike(self, patterns):
```
Does this API have a test in this PR?

cpcloud commented:
Merging on green!

xmnlab commented:
@cpcloud thank you so much for your attention and support!

cpcloud commented:
nice! bombs away!
